File: src/agents/alerter/src/message.js
'use strict';

const ALERT_TITLES = {
  pi:         'Pi offline',
  sensor:     'Sensor ERROR',
  rh:         'RH out of band',
  humidifier: 'Humidifier stuck',
  sht30:      'Primary Humidity Sensor offline',
  scd41:      'CO2 Sensor offline',
};

// Round a number to 1 decimal and strip trailing ".0".
// e.g. 94.39994 -> "94.4", 90 -> "90", 1.5000000000000013 -> "1.5".
// Null/undefined/NaN render as "?" so farmer-facing strings never expose raw
// "null"/"undefined" when upstream state hasn't populated yet.
function fmtNum(n) {
  if (n == null || Number.isNaN(Number(n))) return '?';
  return String(+Number(n).toFixed(1));
}

/**
 * Format elapsed milliseconds as "Xm YYs" (or "Xh YYm" for >= 60 min).
 */
function fmtDuration(ms) {
  const totalSec = Math.round(ms / 1000);
  const totalMin = Math.floor(totalSec / 60);
  const sec = totalSec % 60;
  if (totalMin < 60) {
    return `${totalMin}m ${String(sec).padStart(2, '0')}s`;
  }
  const hours = Math.floor(totalMin / 60);
  const min = totalMin % 60;
  return `${hours}h ${String(min).padStart(2, '0')}m`;
}

/**
 * Format a relative time from a past timestamp to now.
 * e.g. "12m ago" or "5s ago"
 */
function fmtRelative(pastMs, nowMs) {
  const diffMs = nowMs - pastMs;
  const diffSec = Math.round(diffMs / 1000);
  if (diffSec < 60) return `${diffSec}s ago`;
  const diffMin = Math.floor(diffSec / 60);
  return `${diffMin}m ago`;
}

/**
 * formatProblem({alertType, severity, fields, config, nowMs}) -> string
 *
 * Produces a PROBLEM message. Every template includes `config.dashboardUrl` exactly once.
 *
 * fields per alertType:
 *   rh:         { value: number, firstOobMs: number }
 *   pi:         { lastSeenMs: number }
 *   sensor:     { message: string }
 *   humidifier: { onSinceMs: number, rhAtOn: number, currentRh: number }
 */
function formatProblem({ alertType, severity, fields, config, nowMs }) {
  const title = ALERT_TITLES[alertType] || alertType;
  let body = `[PROBLEM · ${severity}] FC-1 · ${title}\n`;

  if (alertType === 'rh') {
    const { value, firstOobMs } = fields;
    body += `Now: ${fmtNum(value)}% · target ${fmtNum(config.rhTarget)}±${fmtNum(config.rhBand)}%\n`;
    if (firstOobMs != null) {
      body += `First OOB: ${fmtRelative(firstOobMs, nowMs)}\n`;
    }
  } else if (alertType === 'pi') {
    const { lastSeenMs, lastKnown } = fields;
    if (lastSeenMs != null) {
      body += `Last seen: ${fmtRelative(lastSeenMs, nowMs)}\n`;
    }
    if (lastKnown) {
      // Phase 29 / 999.39 — situational context for offline alarms.
      // Schema: { rh: number, temp: number, humidifier: 'ON'|'OFF', tsMs: number|null }
      body += `Last sample: RH ${fmtNum(lastKnown.rh)}% · T ${fmtNum(lastKnown.temp)}°C · humidifier ${lastKnown.humidifier}\n`;
      if (lastKnown.tsMs != null) {
        body += `(captured ${fmtRelative(lastKnown.tsMs, nowMs)})\n`;
      }
    }
  } else if (alertType === 'sht30' || alertType === 'scd41') {
    // Hidden 2026-04-25 pending backlog 999.18: lastSeenMs is bootstrapped from
    // alerter boot, not the actual sensor outage onset, so any number we'd print
    // is misleading. Suppress until fc_controller publishes a true wall-clock
    // last-fresh timestamp we can read from.
  } else if (alertType === 'sensor') {
    if (fields && fields.message) {
      body += `${fields.message}\n`;
    }
  } else if (alertType === 'humidifier') {
    const { onSinceMs, rhAtOn, currentRh } = fields || {};
    if (onSinceMs != null) {
      body += `On for: ${fmtDuration(nowMs - onSinceMs)}\n`;
    }
    if (rhAtOn != null && currentRh != null) {
      body += `RH at ON: ${fmtNum(rhAtOn)}% · Now: ${fmtNum(currentRh)}%\n`;
    }
  }

  body += `Open: ${config.dashboardUrl}`;
  return body;
}

/**
 * formatRecovery({alertType, fields, durationMs, config}) -> string
 *
 * fields per alertType:
 *   rh:         { value: number }
 *   pi:         {}
 *   sensor:     {}
 *   humidifier: {}
 */
function formatRecovery({ alertType, fields, durationMs, config }) {
  const title = ALERT_TITLES[alertType] || alertType;
  let body = `[RECOVERY] FC-1 · ${title} back\n`;

  if (alertType === 'rh' && fields && fields.value != null) {
    body += `Now: ${fmtNum(fields.value)}%\n`;
  }

  if (durationMs != null) {
    body += `Was OOB for ${fmtDuration(durationMs)}\n`;
  }

  body += `Open: ${config.dashboardUrl}`;
  return body;
}

/**
 * formatHeartbeat({summary, config, nowMs}) -> string
 *
 * summary: { rh, temp, co2, humidifier, humidifierCycles, piLastSeenSec }
 */
function formatHeartbeat({ summary, config, nowMs: _nowMs }) {
  const { rh, temp, co2, humidifier, humidifierCycles, piLastSeenSec } = summary;
  let body = '[HEARTBEAT] FC-1 watchdog alive\n';
  body += `RH: ${fmtNum(rh)}%  ·  Temp: ${fmtNum(temp)}°C  ·  CO2: ${co2 == null ? '?' : co2} ppm\n`;
  body += `Humidifier: ${humidifier} (cycled ${humidifierCycles}× in last 24h)\n`;
  if (piLastSeenSec != null) {
    body += `Pi last seen: ${piLastSeenSec} seconds ago\n`;
  }
  body += `Open: ${config.dashboardUrl}`;
  return body;
}

module.exports = { formatProblem, formatRecovery, formatHeartbeat, fmtNum, fmtDuration, fmtRelative };

File: src/farm-agent/farm_agent/chamber/message.py
"""
chamber/message.py -- farmer-facing message formatting. Port of message.js.

CHM-02 / D-04: every timestamp renders through ZoneInfo(config.timezone), which
defaults to America/Montevideo. The Node original called
`new Date(tsMs).toISOString().slice(11,16)` (message.js:50-52) -- always UTC --
while config.timezone sat unused and defaulted to America/Toronto anyway. Both
halves were wrong; both are fixed. This is a PRE-DECLARED intentional parity
delta for Phase 64.

Farmer-facing string rules: no em-dashes (feedback_no_em_dashes_in_artifacts),
1-decimal rounding (feedback_round_farmer_numbers), never a raw None.
"""

from __future__ import annotations

import math
from datetime import datetime
from zoneinfo import ZoneInfo

# message.js:3-10
ALERT_TITLES = {
    "pi": "Pi offline",
    "sensor": "Sensor ERROR",
    "rh": "RH out of band",
    "humidifier": "Humidifier stuck",
    "sht30": "Primary Humidity Sensor offline",
    "scd41": "CO2 Sensor offline",
}


def _js_round(x: float) -> int:
    """JS Math.round: half-UP. Python's round() is banker's (round(2.5) == 2)."""
    return math.floor(x + 0.5)


def fmt_num(n) -> str:
    """Round to 1 decimal and strip a trailing '.0'. Port of message.js:16-19.

    - None / NaN -> '?' so a farmer never sees 'null' or 'undefined'
    - 94.39994 -> '94.4'
    - 90       -> '90'
    - 1.5000000000000013 -> '1.5'

    Negative-zero note: Node evaluates String(+(-0.04).toFixed(1)); the unary +
    yields -0 and JS String(-0) is "0", so -0.04 renders "0" (NOT "-0"). Python's
    int(-0.0) is 0, so str(int(r)) reproduces that for free. Verified against node
    2026-07-25. -0.06 still renders "-0.1" -- the sign survives when the magnitude does.
    """
    if n is None:
        return "?"
    try:
        v = float(n)
    except (TypeError, ValueError):
        return "?"
    if math.isnan(v):
        return "?"
    r = round(v, 1)
    if r == int(r):
        return str(int(r))
    return str(r)


def fmt_duration(ms: float) -> str:
    """Elapsed ms as 'Xm YYs', or 'Xh YYm' at >= 60 min. Port of message.js:24-34.

    Example: 6_000 -> '0m 06s'; 95 min -> '1h 35m'.
    """
    total_sec = _js_round(ms / 1000)
    total_min = total_sec // 60
    sec = total_sec % 60
    if total_min < 60:
        return f"{total_min}m {sec:02d}s"
    hours = total_min // 60
    minutes = total_min % 60
    return f"{hours}h {minutes:02d}m"


def fmt_relative(past_ms: int, now_ms: int) -> str:
    """'12m ago' / '5s ago'. Port of message.js:40-46."""
    diff_sec = _js_round((now_ms - past_ms) / 1000)
    if diff_sec < 60:
        return f"{diff_sec}s ago"
    return f"{diff_sec // 60}m ago"


def hhmm(ts_ms: int, tz_name: str) -> str:
    """Render a UTC epoch-ms timestamp as zero-padded LOCAL HH:MM (the D-04 fix).

    Port of message.js:50-52, with the bug removed. Every farmer-facing call site
    MUST route through this -- never format via timezone.utc, never isoformat().

    Example: 2026-07-13T23:30:00Z with 'America/Montevideo' -> '20:30'.
    """
    return datetime.fromtimestamp(ts_ms / 1000, tz=ZoneInfo(tz_name)).strftime("%H:%M")


def format_problem(*, alert_type: str, severity: str, fields: dict, config, now_ms: int) -> str:
    """PROBLEM message. Port of message.js:65-111. Includes dashboard_url exactly once."""
    title = ALERT_TITLES.get(alert_type, alert_type)
    body = f"[PROBLEM · {severity}] FC-1 · {title}\n"

    if alert_type == "rh":
        body += (
            f"Now: {fmt_num(fields.get('value'))}% · "
            f"target {fmt_num(config.rh_target)}±{fmt_num(config.rh_band)}%\n"
        )
        if fields.get("first_oob_ms") is not None:
            body += f"First OOB: {fmt_relative(fields['first_oob_ms'], now_ms)}\n"

    elif alert_type == "pi":
        # Phase 46 D-05/D-06: chamber-level framing. The farmer needs "FC-1 is dark,
        # chamber uncontrolled" up front, not a per-sensor reading.
        last_known = fields.get("last_known")
        last_seen_ms = fields.get("last_seen_ms")
        if last_known is not None and last_known.get("ts_ms") is not None:
            age_min = _js_round((now_ms - last_known["ts_ms"]) / 60_000)
        elif last_seen_ms is not None:
            age_min = _js_round((now_ms - last_seen_ms) / 60_000)
        else:
            age_min = None
        age_str = f"{age_min}m" if age_min is not None else "?"
        if last_known is not None:
            body += (
                f"FC-1 offline ?? no telemetry {age_str}. chamber uncontrolled. "
                f"last RH {fmt_num(last_known.get('rh'))}% @ "
                f"{hhmm(last_known['ts_ms'], config.timezone)}.\n"
            )
        else:
            body += (
                f"FC-1 offline ?? no telemetry {age_str}. chamber uncontrolled. "
                "no recent samples.\n"
            )

    elif alert_type in ("sht30", "scd41"):
        # Backlog 999.18: last_seen_ms is bootstrapped from alerter boot, not the real
        # outage onset, so any number printed here would mislead. Deliberately no body
        # line until fc_controller publishes a true wall-clock last-fresh timestamp.
        pass

    elif alert_type == "sensor":
        if fields and fields.get("message"):
            body += f"{fields['message']}\n"

    elif alert_type == "humidifier":
        on_since = (fields or {}).get("on_since_ms")
        rh_at_on = (fields or {}).get("rh_at_on")
        current_rh = (fields or {}).get("current_rh")
        if on_since is not None:
            body += f"On for: {fmt_duration(now_ms - on_since)}\n"
        if rh_at_on is not None and current_rh is not None:
            body += f"RH at ON: {fmt_num(rh_at_on)}% · Now: {fmt_num(current_rh)}%\n"

    return body + f"Open: {config.dashboard_url}"


def format_recovery(*, alert_type: str, fields: dict, duration_ms: int | None, config) -> str:
    """RECOVERY message. Port of message.js:122-136."""
    title = ALERT_TITLES.get(alert_type, alert_type)
    body = f"[RECOVERY] FC-1 · {title} back\n"
    if alert_type == "rh" and fields and fields.get("value") is not None:
        body += f"Now: {fmt_num(fields['value'])}%\n"
    if duration_ms is not None:
        body += f"Was OOB for {fmt_duration(duration_ms)}\n"
    return body + f"Open: {config.dashboard_url}"


def format_heartbeat(*, summary: dict, config, now_ms: int) -> str:
    """Daily HEARTBEAT message. Port of message.js:143-153."""
    co2 = summary.get("co2")
    body = "[HEARTBEAT] FC-1 watchdog alive\n"
    body += (
        f"RH: {fmt_num(summary.get('rh'))}%  ·  "
        f"Temp: {fmt_num(summary.get('temp'))}°C  ·  "
        f"CO2: {'?' if co2 is None else co2} ppm\n"
    )
    body += (
        f"Humidifier: {summary.get('humidifier')} "
        f"(cycled {summary.get('humidifier_cycles')}× in last 24h)\n"
    )
    if summary.get("pi_last_seen_sec") is not None:
        body += f"Pi last seen: {summary['pi_last_seen_sec']} seconds ago\n"
    return body + f"Open: {config.dashboard_url}"
